docs/python/tutorial-fastapi.md

```diff
@@ -0,0 +1,596 @@
+
+You can also set up Pylance to automatically add imports by looking for the "Auto Import Completions" setting in the Settings UI page (`kb(workbench.action.openSettings)`) and enabling it.
+
+We now have a Redis client object that connects to a Redis server running on the local host (`host="0.0.0.0"`) and listening on port 6379 (`port=6379`). The `db` parameter specifies the Redis database to use. Redis supports multiple databases, and in this code we’re setting to use database 0, which is the default database. We're also passing `decode_responses=True` for the responses to be decoded as strings (instead of bytes).
+
+This client can now be used to perform operations such as getting, setting, and deleting objects in the Redis database.
+
+Let's do some more replacements in our first route `add_item`. Instead of looking into all the keys from our dictionary to find the item name that has been provided, we can fetch that information directly from a redis hash. Let’s assume we will create a hash called `item_name_to_id`, with item names mapped into their IDs once they are added in our app. We can then get the id of the item name we’re receiving by invoking the `hget` method from redis.
+
+3. Delete line 17 (TODO: verify line number):
+    ```python
+    items_names = list(grocery_list.values())
+    ```
+    And replace it with:
+    ```python
+      item_id = redis_client.hget("item_name_to_id", item.item_name)
+     ```
+
+Note that Pylance raises a problem with this change. This is because the `hget` method returns either `str`, or `None` (if the item doesn’t exist). However, the lines below our code that we haven’t replaced yet expect `item_id` to be of type `int`. Let’s address it by renaming the `item_id` symbol.
+
+4. Rename `item_id` with `item_id_str` by selecting the symbol and running the **Rename Symbol** command (`kb(editor.action.rename)`).
+
+    This will rename all occurrences of the symbol in the file, including the one in the `return` statement.
+
+5. Pylance should show a variable type hint next to `item_id_str`. Double click to accept it:
+    ![Variable type hint displayed next to the item_id_str variable.](images/fastapi-tutorial/pylance_redis_typehint.png)
+
+6. If the item doesn't exist, then `item_id_str` will be None. So now we can delete line 18 (TODO: verify line):
+    ```python
+    if item.item_name in items_names:
+    ```
+    And replace it with:
+    ```python
+    if item_id_str is not None:
+    ```
+Now that we have the `item_id` in a string, we just need to convert it to an `int` and add the quantity provided to the item. Because the redis hash we have so far only maps item names to their IDs, we now need to to map item IDs to their names and quantity. One way to do that is to create hashes for each item, in the format `"item_id:{item_id}"`, and provide it with "name" and "quantity" fields.
```

Comment:
> One way to do that is to create hashes for each item, in the format `"item_id:{item_id}"`, and provide it with "name" and "quantity" fields.

I'm having trouble understanding what you're proposing here. Provide what with name and quantity fields?

Reply:
Does `One way to do that is to create hashes for each item, in the format `"item_id:{item_id}"`, and add "name" and "quantity" fields to it.` make it clearer?

Reply:
Maybe the core issue is that I'm not familiar with redis's usage of the word "hash". Is this referring to a hash table where the `item_id` is the key and the `name`/`quantity` are fields on the value? To me "create hashes for each item" sounds like we're running them through a hashing function.

Reply:
that's a good point. I added a paragraph explaining what Redis hashes are in the beginning:
> In Redis, a hash is a data structure that can store multiple key-value pairs. Let's assue we'll create a hash called `item_name_to_id` to store item names (keys) mapped into their IDs (values) once they are added in the app. We can then get the ID of the item name we’re receiving by invoking the `hget` method from Redis.

 and then tweaked this paragraph to try to clarify a bit:
> Now that we have the item ID as a string, we need to convert it to an `int` and update the quantity for the item. Currently, our Redis hash only maps item names to their IDs. To also map item IDs to their names and quantities, we can create a separate Redis hash for each item. We'll use the format `"item_id:{item_id}"` as our hash name to make retrieval by ID easier, and add "name" and "quantity" fields for each hash.

Reply:
Added more explanations throughout the redis section trying to explain better the hashes, how they're structured and why we are using more than one hash (https://github.com/microsoft/vscode-docs/commit/5b8cdfcf5f2f647ca92188167b6efdb98f657a4a). Also added more explanations on the replacements to try to explain better what Redis is doing along the way.